src/python/WMCore/DataStructs/Run.py

```diff
@@ -143,8 +146,14 @@ def __hash__(self):
         """
         Calculate the value of the hash
         """
-        value = self.run.__hash__()
-        value += hash(frozenset(listitems(self.eventsPerLumi)))  # Hash that represents the dictionary
+        if isinstance(self.run, (newstr, newbytes)):
+            value = encodeUnicodeToBytesConditional(self.run, condition=PY3)
+        else:
+            value = encodeUnicodeToBytesConditional(str(self.run), condition=PY3)
+        value = int(hashlib.sha1(value).hexdigest()[:8], 36)
+        # frozenset ensures the same order in python2 and python3, even though it's unordered(?)
+        frozenEvents = encodeUnicodeToBytesConditional(str(frozenset(listitems(self.eventsPerLumi))), condition=PY3)
+        value += int(hashlib.sha1(frozenEvents).hexdigest()[:8], 36)
         return value
```

Comment:
All this complicated code to make sure that the hashid is the same between python2 and python3 (and different executions of the python interpreter).

In addition to that, we need to avoid two exceptions in Python3, they are:
* `TypeError: Unicode-objects must be encoded before hashing`
* `TypeError: object supporting the buffer API required`. A set/frozenset cannot be passed as input value.

There is still 2 problems with the unit tests, because sometimes run (or lumi) can be either an integer or a long object, thus changing the final outcome of this hash function. I still do not know how to gracefully deal with those, besides making sure that we are always dealing with python3 `int` data object.
This is one of the unit tests that fail with the current code changes:
```
WMComponent_t.DBS3Buffer_t.DBSBufferFile_t.DBSBufferFileTest:testAddRunSet changed from success to failure
```

@mapellidario I do not want you to investigate this issue, but in case you have any suggestion from the top of your mind, I'm all ears! ;) (including suggestions to make the current code better!)

Reply:
Hello! I had some fun this morning. https://github.com/dmwm/WMCore/compare/93a110619...mapellidario:10605-suggestion (i left some commented lines in the code to ease reproducibility of the benchmarks)

Some considerations
- i think we were suing `frozenset` because a list is not hash-able because it's mutable, so you can not do `hash(mylist)`. What we did was converting the list into a frozen set (which is an immutable set), then take the hash of that.
- if we want to have the same hash between py2 and py3, then converting the dict `Run.eventsPerLumi` to a string seems like a reasonable option, but we need to make sure that the string representation is the same in py2 and py3. If we do, we fix `test/python/WMComponent_t/DBS3Buffer_t/DBSBufferFile_t.py:DBSBufferFileTest.testAddRunSet`
  1. we need to make sure that the pairs (lumisection, events) are sorted, and we can sort on the lumisection number
  2. we need to remove the `L` from the string in py2
- We can do a similar thing for `self.run`

There is a catch: our new approach that aims at being consistent among py2 and py3 takes about 5 times longer. 

Table: how long (in seconds) it takes to hash the run `run1 = Run(1, 1, 2, 3)` 100'000 times with old approach (`hash()`) and with our new approach (`hashlib.sha1`)
|        | hash() | hashlib.sha1() |
|-|-|-|
| py2 | 0.17    | 1.22      |
| py3 | 0.20    | 1.03     |

Reply:
Thank you very much for this investigation and performance numbers, Dario.

I actually found a way not to return `long` types anymore from the MySQL database. The recommended way, by passing a new `conv=dict` argument when creating the database connection - via `connect` method, did not work. I did not manage to find a way to override the converters when using SQLAlchemy.

However, what I managed to test and got successful results was a patch to the MySQL converters module - in the `py2-mysqldb.spec` here:
https://github.com/farcepest/MySQLdb1/blob/MySQLdb-1.2.4b4/MySQLdb/converters.py#L127

with the following changes:
```
[dmwm@9a960e526008 ~]$ diff -rup converters.py converters.py.new 
--- converters.py	2021-06-18 15:50:20.548763000 +0200
+++ converters.py.new	2021-06-18 15:53:01.234707000 +0200
@@ -143,12 +143,12 @@ conversions = {
     set: Set2Str,
     FIELD_TYPE.TINY: int,
     FIELD_TYPE.SHORT: int,
-    FIELD_TYPE.LONG: long,
+    FIELD_TYPE.LONG: int,
     FIELD_TYPE.FLOAT: float,
     FIELD_TYPE.DOUBLE: float,
     FIELD_TYPE.DECIMAL: float,
     FIELD_TYPE.NEWDECIMAL: float,
-    FIELD_TYPE.LONGLONG: long,
+    FIELD_TYPE.LONGLONG: int,
     FIELD_TYPE.INT24: int,
     FIELD_TYPE.YEAR: int,
     FIELD_TYPE.SET: Str2Set,
```

This is what gives us the best performance, since we do not need to cast data that has already been casted by the database driver.

Python2 int is very large, and I do not think we would have issues with runs/lumis/events numbers too large (we limit them at different places).:
```
(PY2) alannewmacpro:cmsdist amaltaro$ python -c 'import sys; print(sys.maxint)'
9223372036854775807
```

It looks almost like an "extreme" solution to me :), but WMAgent (and perhaps T0) seem to be the only current clients of the MySQLdb py2 library. And, soon'ish we won't even be using it. Please let me know what your thoughts are.

Reply:
That's great news! Every time we avoid playing with string I am happy.

In any case, I would explicitely use `sorted` instead of `frozenset`, since it's not clear to me how reporoducible `frozenset()` ordering is. If i have to guess, being a set, it' likely not reproducible, or is likely to change in the future.

And I would also add something similar to the unittest in my suggestion branch, to check that we are consistently getting the hash that we expect

Reply:
I see. So we would sort the list/set and cast it to string I guess (making it immutable). Is that what you wanted to say?

Sure, please keep your unit test there and I will take it once I test a new build of py2-mysqldb.

Reply:
Yes! That's correct, and I will try to articulate why.

If we use frozenset

```plaintext
frozenEvents = 
encodeUnicodeToBytesConditional( # if it is py3, convert the string to bytes
  str(                                             # convert the frozenset into its string representation
    frozenset(                                # convert the list into an immutable frozenset
      listitems(self.eventsPerLumi) # from the dict, return a list of tuples (lumisection, events)
    )
  )
, condition=PY3) 
```

With this, we are casting the data to a string, but since frozenset is a set and has no defined order, we are not guaranteed which tuple `(lumisection, events)` will come first. If we have a run `run1 = Run(1, 1, 2, 3)`, then it is possible that one time we get `[(1: None), (2: None), (3: None)]`, but we can get as well `[(3: None), (1: None), (2: None)]`. Using the frozenset and checking the hash, we always get different hashes for the same `run1` [1], indeed.

If we replace the `frozenset` with `sorted` we have reproducible results.

```plaintext
frozenEvents = 
encodeUnicodeToBytesConditional( # if it is py3, convert the string to bytes
  str(                                             # convert the frozenset into its string representation
    sorted(                                     # returns a sorted list
      listitems(self.eventsPerLumi) # from the dict, return a list of tuples (lumisection, events)
      , key=lambda x: x[0]               # sorts the list of tuples based on the first item, the lumisection
    )
  )
, condition=PY3) 
```

[1]: 

```plaintext

[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 969116890346 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(2, None), (3, None), (1, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 357614472732 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(3, None), (1, None), (2, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 969116890346 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(2, None), (3, None), (1, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 969116890346 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(2, None), (3, None), (1, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 969116890346 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(2, None), (3, None), (1, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 969116890346 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(2, None), (3, None), (1, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 357614472732 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(3, None), (1, None), (2, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
[dmwm@028e0d827607 WMCore]$ n test/python/WMCore_t/DataStructs_t/Run_t.py:RunTest.testRunHash
F
======================================================================
FAIL: testRunHash (WMCore_t.DataStructs_t.Run_t.RunTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/DataStructs_t/Run_t.py", line 207, in testRunHash
    self.assertEqual(hash(run1), 651310213912)
AssertionError: 969116890346 != 651310213912
-------------------- >> begin captured stdout << ---------------------
frozenset({(2, None), (3, None), (1, None)})

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)

```

Reply:
After a tough debugging and being lost for a while, I figured that I am still getting wrong results here and there because I am misusing these functions. Here is an example:
```
In [4]: int("a2b"[:2], base=16)
Out[4]: 162
```

So, we cannot get the first 16 chars and expect to have an integer of length 16 (very very silly me!). I think I have found a way out, and I will make sure there is a clear comment in the code as well.